concordium-smart-contract-testing/src/invocation/impls.rs

```diff
@@ -0,0 +1,1468 @@
+                            );
+
+                            let success = res.is_success();
+
+                            // Remove the last state changes if the invocation failed.
+                            let state_changed = if !success {
+                                self.invocation_handler.rollback();
+                                false // We rolled back, so no changes were made
+                                      // to this contract.
+                            } else {
+                                let mod_idx_after_invoke =
+                                    self.invocation_handler.modification_index(self.address);
+                                let state_changed = mod_idx_after_invoke != mod_idx_before_invoke;
+                                if state_changed {
+                                    // Update the state field with the newest value from the
+                                    // changeset.
+                                    self.state =
+                                        self.invocation_handler.contract_state(self.address);
+                                }
+                                state_changed
+                            };
+
+                            // Add resume event
+                            let resume_event = ChainEvent::Resumed {
+                                address: self.address,
+                                success,
+                            };
+
+                            self.chain_events.push(resume_event);
+
+                            let resume_res = run_interpreter(
+                                from_interpreter_energy(InterpreterEnergy::from(
+                                    res.remaining_energy,
+                                )),
+                                |energy| {
+                                    v1::resume_receive(
+                                        config,
+                                        res.invoke_response,
+                                        energy,
+                                        &mut self.state,
+                                        state_changed,
+                                        v1::trie::Loader::new(&[][..]),
+                                    )
+                                },
+                            );
+
+                            self.process(resume_res)
+                        }
+                        v1::Interrupt::Upgrade { module_ref } => {
+                            // Add the interrupt event.
+                            self.chain_events.push(interrupt_event);
+
+                            let mut remaining_energy =
+                                from_interpreter_energy(InterpreterEnergy::from(remaining_energy));
+
+                            // Charge a base cost.
+                            remaining_energy = remaining_energy
+                                .saturating_sub(constants::INITIALIZE_CONTRACT_INSTANCE_BASE_COST);
```

Comment:
Use the suggested `tick` here and elswhere where doing this `x = x.saturating_sub`